### PR TITLE
Copy the old implementation regularized van Genuchten to MPL with some improvements

### PR DESCRIPTION
**Pull request by wenqing**
as titled.
In the old  old implementation, pc of  S  in [Smax, 1] is wrong, which should be zero. 

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.3.1)
2. [x] Tests covering your feature were added? Yes.
The following figure shows the curve of the regularized pc(S) in the test:
RegularizedVanGenuchten
 

3. [x] Any new feature or behavior change was documented? Yes

### CODE REVIEW

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    if (Sg < Sg_r_)
+    {
+        return 0.0;
+    }
+
+    return getPcBarvGSg(Sg_max_) + getdPcdSvGBar(Sg_max_) * (Sg - Sg_max_);
+}
+
+PropertyDataType CapillaryPressureRegularizedVanGenuchten::dValue(
+    VariableArray const& variable_array, Variable const /*primary_variable*/,
```
Comment (endJunction) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:72]: primary variable must be checked for being saturation.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#include "MaterialLib/MPL/VariableType.h"
+#include "MaterialLib/MPL/Utils/CheckVanGenuchtenExponentRange.h"
+
+namespace MaterialPropertyLib
+{
+void checkSaturationRange(const double Sl)
+{
+    if (Sl < 0 || Sl > 1)
+    {
+        OGS_FATAL("The saturation of {:g} is  out of its range of [0, 1]", Sl);
```
Comment (endJunction) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:28]: Error message uses closed interval, but check is for open interval.
Minor: Extra space between "is" and "out".

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    }
+}
+
+CapillaryPressureRegularizedVanGenuchten::
+    CapillaryPressureRegularizedVanGenuchten(
+        double const residual_liquid_saturation,
+        double const maximum_liquid_saturation,
+        double const exponent,
+        double const p_b)
+    : Sl_r_(residual_liquid_saturation),
```
Comment (endJunction) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:38]: No need to store the value if it is only used in the ctor.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    if (!(Sg < Sg_r_ || Sg > Sg_max_))
+    {
+        return getPcBarvGSg(Sg);
+    }
+    if (Sg < Sg_r_)
+    {
+        return 0.0;
+    }
+
+    return getPcBarvGSg(Sg_max_) + getdPcdSvGBar(Sg_max_) * (Sg - Sg_max_);
```
Comment (endJunction) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:68]: Rather expensive computations for two constants. Could be saved upon creation instead of the unused variables.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 1 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: added.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 2 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: S>1 is not allowed.
Extra space between "is" and "out" was removed.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 3 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: Removed.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 4 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: Added two constant members for that.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    }
+}
+
+CapillaryPressureRegularizedVanGenuchten::
+    CapillaryPressureRegularizedVanGenuchten(
+        double const residual_liquid_saturation,
+        double const maximum_liquid_saturation,
+        double const exponent,
+        double const p_b)
+    : Sl_max_(maximum_liquid_saturation),
```
Comment (endJunction) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:38]: Why not remove this too?

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 2 above)
Comment (TomFischer) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: The if condition in line 26 handles unallowed input. The error message corresponding to an input value 1 + 1e-8 would be:
``'The saturation of 1 is out of range of [0, 1].```
Some users may not be able to see the error in it.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+double CapillaryPressureRegularizedVanGenuchten::getdPcdSvG(
+    const double Sg) const
+{
+    double const n = 1 / (1 - m_);
+    double const Se = (Sg_max_ - Sg) / (Sg_max_ - Sg_r_);
+    return p_b_ * (1 / (m_ * n)) * (1 / (Sg_max_ - Sg_r_)) *
+           std::pow(std::pow(Se, (-1 / m_)) - 1, (1 / n) - 1) *
+           std::pow(Se, (-1 / m_)) / Se;
```
Comment (TomFischer) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:134]:
```suggestion
   auto const t = std::pow(Se, (-1 / m_);
    return p_b_ * (1 / (m_ * n)) * (1 / (Sg_max_ - Sg_r_)) *
           std::pow(t - 1, - m_) *
           t / Se;
```

Comment (TomFischer) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:134]: I'm not sure if the variable `n` is necessary.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.h [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        double const exponent,
+        double const p_b);
+
+    void checkScale() const override
+    {
+        if (!std::holds_alternative<Medium*>(scale_))
+        {
+            OGS_FATAL(
+                "The property 'CapillaryPressureRegularizedVanGenuchten' is "
+                "implemented on the "
```
Comment (TomFischer) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.h:45]: Message would fit in two lines.

File: MaterialLib/MPL/Utils/CheckVanGenuchtenExponentRange.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "CheckVanGenuchtenExponentRange.h"
+
+#include "BaseLib/Error.h"
+
+namespace MaterialPropertyLib
+{
+void checkVanGenuchtenExponentRange(const double m)
+{
+    if (!(m > 0 && m < 1))
```
Comment (TomFischer) [inline MaterialLib/MPL/Utils/CheckVanGenuchtenExponentRange.cpp:20]: IMHO the following version is more readable.
```suggestion
    if (m <= 0 || m >= 1))
```

File: Tests/MaterialLib/TestMPLCapillaryPressureRegularizedVanGenuchten.cpp [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    double const t = std::numeric_limits<double>::quiet_NaN();
+    double const dt = std::numeric_limits<double>::quiet_NaN();
+
+    std::array<double, 31> S{
+        {0,        0.0333333, 0.0666667, 0.1,      0.133333, 0.166667, 0.2,
+         0.233333, 0.266667,  0.3,       0.333333, 0.366667, 0.4,      0.433333,
+         0.466667, 0.5,       0.533333,  0.566667, 0.6,      0.633333, 0.666667,
+         0.7,      0.733333,  0.766667,  0.8,      0.833333, 0.866667, 0.9,
+         0.933333, 0.966667,  1}};
+    // Data for S in [0, S_max] calculated by from the old implementation.
```
Comment (TomFischer) [inline Tests/MaterialLib/TestMPLCapillaryPressureRegularizedVanGenuchten.cpp:78]:
```suggestion
    // Data for S in [0, S_max] calculated by the old implementation.
```

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 2 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: Changed the output format to {:e}

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 5 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: Removed.

**Comment by wenqing** [comment]
> Technically OK.
> 
> Naming of variables is not consistent, _e.g._ there are two ways how liquid saturation is called `S` and `Sl`, and two other ways how gas saturation is called, `S_g` and `Sg`. This is again different from the existing implementations using `S_L`, for example.

Use Sl and Sg in this class.
> 
> Testing the derivative against numerical one (central differences, eps=1e-8) is showing bigger error in the 0 to 0.1 region:
> This is to be improved or explained.

In that range, dpc/dS is constant but with large magnitude as
```
   calculated       analytic
 -5.36458e+12 -5.36458e+12
-5.36458e+12  -5.36458e+12
```

File: Tests/MaterialLib/TestMPLCapillaryPressureRegularizedVanGenuchten.cpp (hunk 9 above)
Comment (wenqing) [reply Tests/MaterialLib/TestMPLCapillaryPressureRegularizedVanGenuchten.cpp]: Changed.

File: MaterialLib/MPL/Utils/CheckVanGenuchtenExponentRange.cpp (hunk 8 above)
Comment (wenqing) [reply MaterialLib/MPL/Utils/CheckVanGenuchtenExponentRange.cpp]: Changed.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.h (hunk 7 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.h]: Changed.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp (hunk 6 above)
Comment (wenqing) [reply MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp]: Changed with the suggestion but kept `n`.

**Comment by wenqing** [comment]
Since every commit is small. The changes for reply were made by amending the commits.

**Comment by wenqing** [comment]
Citation: The old citation in the old implementation does not explicitly provide the information of the model. I replaced it with the paper by @Yonghui56 and @HBShaoUFZ, in which the regularization has a slightly different form. The  implementation actually provides a linear extension.

**Comment by endJunction** [comment]
> > Testing the derivative against numerical one (central differences, eps=1e-8) is showing bigger error in the 0 to 0.1 region:
> > This is to be improved or explained.
> 
> In that range, dpc/dS is constant but with large magnitude as
> 
> ```
>    calculated       analytic
>  -5.36458e+12 -5.36458e+12
> -5.36458e+12  -5.36458e+12
> ```
How does the large magnitude explain the big deviation from the numerical derivative?

**Comment by wenqing** [comment]
> > > Testing the derivative against numerical one (central differences, eps=1e-8) is showing bigger error in the 0 to 0.1 region:
> > > This is to be improved or explained.
> > 
> > 
> > In that range, dpc/dS is constant but with large magnitude as
> > ```
> >    calculated       analytic
> >  -5.36458e+12 -5.36458e+12
> > -5.36458e+12  -5.36458e+12
> > ```
> 
> How does the large magnitude explain the big deviation from the numerical derivative?

It is because of the truncation error in the numerical calculation of the derivative of that straight segment of the curve. For that case, larger dS gives smaller error. Here are a test for that
```
dS = 1.e-7  
0.5*( dPcdSvGBarSg_max_ * (1-0.0333333-1.e-7 - Sg_max_)-( dPcdSvGBarSg_max_ * (1-0.0333333+1.e-7 - Sg_max_)))/1.e-7  --> error: 233.

dS = 1.
0.5*( dPcdSvGBarSg_max_ * (1-0.0333333-1. - Sg_max_)-( dPcdSvGBarSg_max_ * (1-0.0333333+1.0 - Sg_max_)))/1.  --> error: 0.

```
Not set the individual dS for [0,Sr] and [Sr,1] in the test.

File: MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    checkSaturationRange(Sl);
+
+    double const Sg = 1 - Sl;
+    if (!(Sg < Sg_r_ || Sg > Sg_max_))
+    {
+        return getPcBarvGSg(Sg);
+    }
+    if (Sg < Sg_r_)
+    {
+        return 0.0;
```
Comment (wenqing) [inline MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureRegularizedVanGenuchten.cpp:65]: Note: the old implementation gives a non-zero  wrong extension here. The correction sightly influences the results of the corresponding benchmark, which will be updated in another PR for using this new MPL implementation.
